plugin/hosts/hosts.go

```diff
@@ -52,15 +52,15 @@ func (h Hosts) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (
 		answers = aaaa(qname, h.options.ttl, ips)
 	}
 
-	if len(answers) == 0 {
+	// Only on NXDOMAIN we will fallthrough.
+	if len(answers) == 0 && !h.otherRecordsExist(qname) {
 		if h.Fall.Through(qname) {
 			return plugin.NextOrFailure(h.Name(), h.Next, ctx, w, r)
 		}
+
 		// We want to send an NXDOMAIN, but because of /etc/hosts' setup we don't have a SOA, so we make it REFUSED
 		// to at least give an answer back to signals we're having problems resolving this.
-		if !h.otherRecordsExist(qname) {
-			return dns.RcodeServerFailure, nil
-		}
+		return dns.RcodeRefused, nil
```

Comment:
I meant changing the comment, not the return code itself

Reply:
also making this change without failing tests means this is lacking coverage. As you are touching this code; please also add a test for this :)

Reply:
OK, I'm going to add some new tests which will cover this line and my new code!